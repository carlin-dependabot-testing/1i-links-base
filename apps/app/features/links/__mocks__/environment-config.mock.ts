import { type EnvironmentConfig } from '@/features/links/types/environment.types'

export const mockEnvironmentConfig: EnvironmentConfig = {
  environmentColors: {
    production: '#0f0',
    staging: '#ff0',
    integration: '#00f'
  },
  environmentLabels: {
    production: 'Production',
    staging: 'Staging',
    integration: 'Integration'
  },
  enabledEnvironments: ['production', 'staging', 'integration'],
  environmentOrder: ['production', 'staging', 'integration']
}
