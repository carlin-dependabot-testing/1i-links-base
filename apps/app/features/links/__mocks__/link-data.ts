import { type LinksApp } from '@/features/links/types/link.types'

export const mockLinkData: Record<string, LinksApp> = {
  basic: {
    name: 'Test App',
    description: 'Test Description',
    category: 'Category 1',
    color: '#000000',
    environments: {
      production: 'https://test.com'
    },
    tags: ['test-tag'],
    createdAt: '2024-03-20',
    timestamp: '2024-03-20'
  },
  withMultipleEnvironments: {
    name: 'Multi Env App',
    description: 'App with multiple environments',
    category: 'Category 2',
    color: '#000000',
    environments: {
      production: 'https://prod.test.com',
      staging: 'https://staging.test.com',
      integration: 'https://dev.test.com'
    },
    tags: ['test-tag'],
    createdAt: '2024-03-20',
    timestamp: '2024-03-20'
  }
} as const
