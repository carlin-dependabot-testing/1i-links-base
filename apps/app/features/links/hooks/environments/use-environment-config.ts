import { type Environment } from '@/features/environment/types/environment.types'
import { type EnvironmentConfig } from '@/features/links/types/environment.types'
import { useSettings } from '@/features/settings/hooks/use-settings'

export const useEnvironmentConfig = (): EnvironmentConfig => {
  const { settings } = useSettings()

  if (!settings) {
    return {
      enabledEnvironments: [] as Environment[],
      environmentLabels: {} as Record<Environment, string>,
      environmentColors: {} as Record<Environment, string>,
      environmentOrder: [] as Environment[]
    }
  }

  const defaultLabels: Record<Environment, string> = {
    production: 'Production',
    staging: 'Staging',
    integration: 'Integration'
  }

  const defaultColors: Record<Environment, string> = {
    production: '#000000',
    staging: '#666666',
    integration: '#999999'
  }

  return {
    enabledEnvironments: settings.environments.configuration.enabled,
    environmentLabels: {
      ...defaultLabels,
      ...settings.environments.display.labels
    },
    environmentColors: {
      ...defaultColors,
      ...settings.environments.display.colors
    },
    environmentOrder: settings.environments.configuration.order
  }
}
