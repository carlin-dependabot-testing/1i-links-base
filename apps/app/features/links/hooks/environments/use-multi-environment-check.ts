import { useMemo } from 'react'

import { useLinks } from '../links/use-links'

/** Checks if there are any multi-environment links in the specified category */
export const useMultiEnvironmentCheck = (categoryId?: string) => {
  const { links } = useLinks()

  const filteredLinks = useMemo(() => {
    if (!categoryId) return links
    return links.filter(link => link.category === categoryId)
  }, [links, categoryId])

  const hasMultipleEnvironments = useMemo(() => {
    return filteredLinks.some(link => {
      const environments = link.environments || {}
      // Check if there are any non-production environments (staging or integration) with valid URLs
      return environments.staging || environments.integration
    })
  }, [filteredLinks])

  return {
    hasMultipleEnvironments
  }
}
