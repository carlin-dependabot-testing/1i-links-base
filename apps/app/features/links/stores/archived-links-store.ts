import { persist } from 'zustand/middleware'
import { type StoreApi, createStore } from 'zustand/vanilla'

export type ArchivedLinksState = {
  archivedLinks: string[]
}

export type ArchivedLinksActions = {
  toggleArchived: (linkName: string) => void
  isArchived: (linkName: string) => boolean
  unarchiveAll: () => void
}

export type ArchivedLinksStore = ArchivedLinksState & ArchivedLinksActions

const defaultInitState: ArchivedLinksState = {
  archivedLinks: []
}

export const createArchivedLinksStore = (
  initState: ArchivedLinksState = defaultInitState
): StoreApi<ArchivedLinksStore> => {
  return createStore<ArchivedLinksStore>()(
    persist(
      (set, get) => ({
        ...initState,
        toggleArchived: (linkName: string) =>
          set(state => {
            const isCurrentlyArchived = state.archivedLinks.includes(linkName)
            return {
              archivedLinks: isCurrentlyArchived
                ? state.archivedLinks.filter(name => name !== linkName)
                : [...state.archivedLinks, linkName]
            }
          }),
        isArchived: (linkName: string) =>
          get().archivedLinks.includes(linkName),
        unarchiveAll: () => set({ archivedLinks: [] })
      }),
      {
        name: 'archived-links-store'
      }
    )
  )
}
