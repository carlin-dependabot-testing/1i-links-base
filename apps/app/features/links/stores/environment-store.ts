import { create } from 'zustand'

import { type Environment } from '@/features/environment/types/environment.types'
import { getSettingsJson } from '@/features/setup/utils/load-json-data'

interface EnvironmentState {
  currentEnvironment: Environment | null
  availableEnvironments: Environment[]
  isLoading: boolean
  error: string | null
  setCurrentEnvironment: (environment: Environment) => void
  initializeEnvironments: () => Promise<void>
}

export const useEnvironmentStore = create<EnvironmentState>(set => ({
  currentEnvironment: null,
  availableEnvironments: [],
  isLoading: true,
  error: null,

  setCurrentEnvironment: environment => {
    set({ currentEnvironment: environment })
  },

  initializeEnvironments: async () => {
    try {
      set({ isLoading: true, error: null })
      const settings = await getSettingsJson()

      if (!settings?.environments?.configuration?.order) {
        set({
          currentEnvironment: null,
          availableEnvironments: [],
          error: 'No environments configured'
        })
        return
      }

      const environments = settings.environments.configuration.order

      set({
        currentEnvironment: environments[0] || null,
        availableEnvironments: environments,
        error: null
      })
    } catch (error) {
      set({
        error: 'Failed to load environments',
        currentEnvironment: null,
        availableEnvironments: []
      })
    } finally {
      set({ isLoading: false })
    }
  }
}))

// Initialize environments when the store is first created
void useEnvironmentStore.getState().initializeEnvironments()
