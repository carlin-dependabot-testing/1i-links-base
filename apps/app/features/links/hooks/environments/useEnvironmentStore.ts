'use client'

import { useEnvironmentStore as baseStore } from '@/features/links/stores/environment-store'

export const useEnvironmentStore = () => {
  const {
    currentEnvironment,
    availableEnvironments,
    setCurrentEnvironment,
    isLoading,
    error
  } = baseStore()

  return {
    currentEnvironment,
    availableEnvironments,
    setCurrentEnvironment,
    isLoading,
    error
  }
}

export default useEnvironmentStore
