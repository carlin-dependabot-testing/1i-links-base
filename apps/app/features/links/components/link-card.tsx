'use client'

import { useEffect, useRef, useState } from 'react'

import { Archive, Info } from 'lucide-react'

import { ImageError } from '@/utils/image-error'

import { BADGE_STYLE, DEFAULT_COLOR } from '@/constants'
import { EnvironmentBadge } from '@/features/environment/components/environment-badge'
import { type Environment } from '@/features/environment/types/environment.types'
import { useFavourites } from '@/features/favourite/hooks/use-favourites'
import { CardMoreMenu } from '@/features/links/components/card-more-menu'
import { LinkDetailsModal } from '@/features/links/components/link-details-modal'
import { useEnvironmentStore } from '@/features/links/hooks/environments/useEnvironmentStore'
import { useLinkClickStore } from '@/features/links/stores/use-link-click-store'
import { useRecentLinksStore } from '@/features/links/stores/use-recent-links-store'
import { type LinksApp } from '@/features/links/types/link.types'
import { useUserSettingsStore } from '@/features/user/stores/useUserSettingsStore'

import { Badge } from '@links-base/ui/badge'
import { CardContent, CardDescription, CardTitle } from '@links-base/ui/card'
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger
} from '@links-base/ui/tooltip'
import { cn } from '@links-base/ui/utils'

interface LinkCardProps {
  app: LinksApp
  handleLinkClick: (url: string) => void
  clickedLinks: string[]
  handleAddToFavorites: () => void
  isArchived?: boolean
  onArchiveToggle: (name: string) => void
  isFavourite: boolean
  className?: string
  onDismissNew?: (appName: string) => void
  showManageOptions?: boolean
}

export const LinkCard = ({
  app,
  handleLinkClick,
  clickedLinks,
  handleAddToFavorites,
  isArchived = false,
  onArchiveToggle,
  className,
  onDismissNew,
  showManageOptions = true
}: LinkCardProps) => {
  const tagsContainerRef = useRef<HTMLDivElement>(null)
  const favourites = useFavourites()
  const isPinned = favourites.isFavourite(app.name)
  const { currentEnvironment } = useEnvironmentStore()

  // Get user settings
  const {
    cardSize,
    cardSizeConfigs,
    showGradients,
    showDescriptions,
    showTags,
    showIcons,
    showEnvironmentBadges
  } = useUserSettingsStore()

  const currentConfig = cardSizeConfigs[cardSize] || cardSizeConfigs.default

  const store = useLinkClickStore()

  const { hasVisitedRecent, recentLinks } = useRecentLinksStore()

  const [showInstructionsModal, setShowInstructionsModal] = useState(false)

  useEffect(() => {
    if (!app.tags?.length) return

    const calculateVisibleTags = () => {
      const container = tagsContainerRef.current
      if (!container) return

      requestAnimationFrame(() => {
        const containerWidth = container.offsetWidth
        let currentWidth = 0

        const badges = container.getElementsByClassName('badge-element')

        for (const badge of badges) {
          currentWidth += (badge as HTMLElement).offsetWidth + 8

          if (currentWidth > containerWidth) {
            break
          }
        }
      })
    }

    calculateVisibleTags()
    window.addEventListener('resize', calculateVisibleTags)

    return () => window.removeEventListener('resize', calculateVisibleTags)
  }, [app.tags])

  const getUrl = () => {
    if (!app.environments) return ''
    return (
      app.environments[currentEnvironment as keyof typeof app.environments] ||
      app.environments.production
    )
  }

  const url = getUrl()

  const hasMultipleEnvironments =
    app.environments &&
    Object.entries(app.environments).filter(
      ([_, value]) => value && value !== ''
    ).length > 1

  const handleClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
    e.preventDefault()
    e.stopPropagation()

    const url = getUrl()
    if (!url) return

    store.incrementClickCount(app.name, {
      url: url,
      environment: currentEnvironment as string
    })

    handleLinkClick(url)

    setTimeout(() => {
      window.open(url, '_blank', 'noopener,noreferrer')
    }, 100)
  }

  // Get the brand color from the app's icon, defaulting to a neutral color if not available
  const brandColor = app?.color || DEFAULT_COLOR

  const gradientStyle = showGradients
    ? ({
        // Light mode gradient
        backgroundImage: `linear-gradient(to bottom left, ${brandColor}35, ${brandColor}15, transparent)`,
        content: '',
        position: 'absolute',
        top: '-3rem',
        right: '-3rem',
        width: '16rem',
        height: '16rem',
        filter: 'blur(12px)',
        zIndex: 0
      } as React.CSSProperties)
    : {}

  const darkGradientStyle = showGradients
    ? ({
        // Dark mode gradient
        backgroundImage: `linear-gradient(to bottom left, ${brandColor}90, ${brandColor}60, transparent)`
      } as React.CSSProperties)
    : {}

  const cardStyle = {
    '--card-shadow-color': `${brandColor}40`
  } as React.CSSProperties

  const shouldShowNewBadge = () => {
    if (hasVisitedRecent) return false
    if (clickedLinks.includes(url)) return false
    return recentLinks.includes(app.name)
  }

  return (
    <div className="relative h-full">
      {shouldShowNewBadge() && (
        <div className={cn(BADGE_STYLE, 'group absolute -left-3 -top-3 z-50')}>
          <span>New</span>
          {onDismissNew && (
            <button
              onClick={e => {
                e.preventDefault()
                e.stopPropagation()
                onDismissNew(app.name)
              }}
              className="ml-1 opacity-0 transition-opacity hover:text-white/80 group-hover:opacity-100"
              aria-label="Dismiss new badge"
            >
              Ã—
            </button>
          )}
        </div>
      )}
      <div
        style={cardStyle}
        className={cn(
          'relative flex flex-none select-none flex-col overflow-hidden rounded-lg border bg-card text-card-foreground shadow-sm hover:shadow-lg dark:shadow-slate-500/50',
          'transition-all duration-300 ease-in-out',
          'hover:shadow-[0_8px_16px_-2px_var(--card-shadow-color)]',
          'w-full',
          isArchived && [
            'border-dashed border-muted-foreground/50',
            'bg-muted/50 dark:bg-muted/20'
          ],
          isPinned && 'opacity-50 transition-opacity hover:opacity-100',
          className
        )}
      >
        <div
          style={gradientStyle}
          className="dark:[&]:!bg-none" // Reset background in dark mode
        />
        {showGradients && (
          <div
            style={darkGradientStyle}
            className="absolute right-[-3rem] top-[-3rem] z-0 hidden h-64 w-64 blur-xl dark:block"
          />
        )}
        {isPinned && (
          <div className="absolute inset-0 z-10 flex items-center justify-center bg-background/50 opacity-0 backdrop-blur-[1px] transition-opacity hover:opacity-100">
            <p className="text-sm font-medium text-muted-foreground">
              Already pinned to favorites
            </p>
          </div>
        )}
        <CardContent
          className={cn(
            'z-10 flex flex-1 flex-col gap-2',
            'transition-all duration-300 ease-in-out',
            currentConfig?.padding || 'p-4',
            cardSize === 'compact' && '!p-2'
          )}
        >
          <div className="flex w-full items-center justify-between">
            <div className="flex items-center gap-2">
              {showIcons && <ImageError app={app} />}
              {cardSize === 'compact' ? (
                <CardTitle className="text-sm font-medium transition-all duration-300 ease-in-out">
                  <a
                    href={url}
                    className="before:absolute before:bottom-0 before:left-0 before:right-0 before:top-0 before:z-10 before:content-['']"
                    onClick={handleClick}
                    target="_blank"
                    rel="noopener noreferrer nofollow"
                  >
                    {app.name}
                  </a>
                </CardTitle>
              ) : null}
              {app.instructions && (
                <div className="relative z-10">
                  <TooltipProvider>
                    <Tooltip>
                      <TooltipTrigger asChild>
                        <Info
                          className="cursor-pointer text-blue-500"
                          size={16}
                          onClick={e => {
                            e.preventDefault()
                            e.stopPropagation()
                            setShowInstructionsModal(true)
                          }}
                        />
                      </TooltipTrigger>
                      <TooltipContent>
                        <p>Click for more information</p>
                      </TooltipContent>
                    </Tooltip>
                  </TooltipProvider>
                </div>
              )}
            </div>
            <div className="flex items-center gap-2">
              <div className="flex items-center gap-2">
                {isArchived && (
                  <div
                    className="z-20 flex items-center gap-1.5 rounded-md border border-muted-foreground/30 bg-background/80 px-2 py-1"
                    aria-label="This link is archived"
                  >
                    <Archive className="h-3.5 w-3.5 text-muted-foreground" />
                    <span className="text-xs font-medium text-muted-foreground">
                      Archived
                    </span>
                  </div>
                )}

                {showEnvironmentBadges &&
                  hasMultipleEnvironments &&
                  !isArchived && (
                    <EnvironmentBadge
                      environment={currentEnvironment as Environment}
                    />
                  )}
              </div>
              <CardMoreMenu
                app={app}
                handleAddToFavorites={handleAddToFavorites}
                isArchived={isArchived}
                onArchiveToggle={() => onArchiveToggle(app.name)}
                onShowAccess={() => {
                  setShowInstructionsModal(true)
                }}
                showManageOptions={showManageOptions}
              />
            </div>
          </div>

          {cardSize !== 'compact' && (
            <div className="flex flex-1 origin-top flex-col transition-all duration-300 ease-in-out">
              <CardTitle className="text-sm font-medium transition-opacity duration-300">
                <a
                  href={url || '#'}
                  className="before:absolute before:bottom-0 before:left-0 before:right-0 before:top-0 before:z-10 before:content-['']"
                  onClick={handleClick}
                  target="_blank"
                  rel="noopener noreferrer nofollow"
                >
                  {app.name}
                </a>
              </CardTitle>
              <div className="flex flex-1 flex-col justify-between">
                {showDescriptions && (
                  <CardDescription
                    className={cn(
                      'text-xs transition-all duration-300 ease-in-out',
                      currentConfig.descriptionClasses
                    )}
                  >
                    {app.description}
                  </CardDescription>
                )}
                {showTags && app.tags && app.tags.length > 0 && (
                  <div className="z-30 mt-auto flex items-center gap-1.5 pt-2">
                    {/* Show primary tag */}
                    <Badge
                      variant="secondary"
                      className="border border-muted bg-background/80 px-1.5 py-0.5 text-[10px]"
                    >
                      {app.tags[0]}
                    </Badge>

                    {/* If there are additional tags, show count in subtle way */}
                    {app.tags.length > 1 && (
                      <span className="text-[10px] text-muted-foreground">
                        +{app.tags.length - 1}
                      </span>
                    )}
                  </div>
                )}
              </div>
            </div>
          )}
        </CardContent>
      </div>

      <LinkDetailsModal
        isOpen={showInstructionsModal}
        onClose={() => setShowInstructionsModal(false)}
        link={app}
      />
    </div>
  )
}
