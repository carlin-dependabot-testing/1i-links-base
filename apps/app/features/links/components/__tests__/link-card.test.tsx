import { render } from '@/utils/test-utils'

import * as environmentHooks from '@/features/links/hooks/environments/use-environment-config'
import { mockUserSettings } from '@/features/user/__mocks__/user-settings.mock'
import * as userSettingsStore from '@/features/user/stores/useUserSettingsStore'

import { mockEnvironmentConfig } from '../../__mocks__/environment-config.mock'
import { mockLinkData } from '../../__mocks__/link-data'
import { LinkCard } from '../link-card'

jest.mock('@/features/links/hooks/environments/use-environment-config', () => ({
  useEnvironmentConfig: jest.fn()
}))

jest.mock('@/features/user/stores/useUserSettingsStore', () => ({
  useUserSettingsStore: jest.fn()
}))

const renderComponent = (props = {}) => {
  const defaultProps = {
    app: mockLinkData.basic,
    handleLinkClick: jest.fn(),
    clickedLinks: [],
    handleAddToFavorites: jest.fn(),
    selectedEnvironment: 'production',
    onArchiveToggle: jest.fn(),
    isFavourite: false
  }

  return render(<LinkCard {...defaultProps} {...props} />)
}

describe('<LinkCard />', () => {
  beforeEach(() => {
    jest
      .spyOn(environmentHooks, 'useEnvironmentConfig')
      .mockReturnValue(mockEnvironmentConfig)
    jest
      .spyOn(userSettingsStore, 'useUserSettingsStore')
      .mockReturnValue(mockUserSettings)

    jest.clearAllMocks()
  })

  it.skip('should render correctly', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  // it('should display the app name and description', () => {
  //   renderComponent();
  //   expect(screen.getByText('Test App')).toBeInTheDocument();
  //   expect(screen.getByText('Test Description')).toBeInTheDocument();
  // });

  // it('should display the app description', () => {
  //   renderComponent();
  //   expect(screen.getByText('Test Description')).toBeInTheDocument();
  // });

  // it('should handle click events', () => {
  //   const handleLinkClick = jest.fn();
  //   renderComponent({ handleLinkClick });

  //   const link = screen.getByRole('link', { name: 'Test App' });
  //   fireEvent.click(link);

  //   expect(handleLinkClick).toHaveBeenCalledWith('https://test.com');
  // });

  // it('should display tags when provided', () => {
  //   renderComponent();
  //   expect(screen.getByText('test-tag')).toBeInTheDocument();
  // });

  // it('should show archived state when isArchived is true', () => {
  //   renderComponent({ isArchived: true });
  //   expect(screen.getByText('Archived')).toBeInTheDocument();
  // });

  // it('should not show archived state when isArchived is true', () => {
  //   renderComponent();
  //   expect(screen.queryByText('Archived')).not.toBeInTheDocument();
  // });

  // it.skip('should show instructions icon when instructions are provided', async () => {
  //   const user = userEvent.setup();

  //   const appWithInstructions = {
  //     ...mockLinkData.basic,
  //     instructions: 'Test instructions'
  //   };

  //   renderComponent({ app: appWithInstructions });

  //   const moreInformationButton = screen.getByRole('button', { name: /Open (more menu|menu)/i });
  //   await user.click(moreInformationButton);

  //   const moreInformationMenuItem = screen.getByRole('menuitem', { name: 'More Information' });
  //   await user.click(moreInformationMenuItem);
  // });

  // it.skip('should show environment badge where if there is an environment', () => {

  //   renderComponent({ app: mockLinkData.withMultipleEnvironments });

  //   // expect(screen.getByText('Production')).toBeInTheDocument();
  // });
})
