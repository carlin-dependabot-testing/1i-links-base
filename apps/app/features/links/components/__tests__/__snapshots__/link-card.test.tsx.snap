// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkCard /> should render correctly 1`] = `
<div>
  <div
    class="group/sidebar-wrapper flex min-h-svh w-full has-[[data-variant=inset]]:bg-sidebar"
    style="--sidebar-width: 16rem; --sidebar-width-icon: 3rem;"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
    </script>
    <div
      class="relative h-full"
    >
      <div
        class="flex-none rounded-lg border bg-card text-card-foreground shadow-sm relative dark:shadow-slate-500/50 select-none flex flex-col overflow-hidden transition-all duration-300 ease-in-out hover:shadow-[0_8px_16px_-2px_var(--card-shadow-color)]"
        style="--card-shadow-color: #00000040;"
      >
        <div
          style="content: ''; position: absolute; top: -3rem; right: -3rem; width: 16rem; height: 16rem; filter: blur(12px); z-index: 0;"
        />
        <div
          class="flex flex-col gap-2 flex-1 z-10 transition-all duration-300 ease-in-out p-4"
        >
          <div
            class="w-full flex justify-between items-center"
          >
            <div
              class="flex items-center gap-2"
            >
              <img
                alt="Favicon"
                class="rounded"
                data-nimg="1"
                decoding="async"
                height="32"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=64&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=32&q=75 1x, /_next/image?url=https%3A%2F%2Fwww.google.com%2Fs2%2Ffavicons%3Fsz%3D32%26domain_url%3Dhttps%253A%252F%252Ftest.com&w=64&q=75 2x"
                style="color: transparent; object-fit: cover;"
                width="32"
              />
            </div>
            <div
              class="flex items-center gap-2"
            >
              <div
                class="flex items-center gap-2"
              />
              <button
                aria-expanded="false"
                aria-haspopup="menu"
                class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent hover:text-accent-foreground h-8 w-8 p-0 z-30"
                data-state="closed"
                id="radix-:r0:"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  Open menu
                </span>
                <svg
                  class="lucide lucide-ellipsis-vertical h-4 w-4"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="1"
                  />
                  <circle
                    cx="12"
                    cy="5"
                    r="1"
                  />
                  <circle
                    cx="12"
                    cy="19"
                    r="1"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="transition-all duration-300 ease-in-out origin-top flex flex-col flex-1"
          >
            <h3
              class="tracking-tight text-sm font-medium transition-opacity duration-300"
            >
              <a
                class="before:absolute before:top-0 before:left-0 before:right-0 before:bottom-0 before:content-[''] before:z-10"
                href="https://test.com"
                rel="noopener noreferrer nofollow"
                target="_blank"
              >
                Test App
              </a>
            </h3>
            <div
              class="flex flex-col flex-1 justify-between"
            >
              <p
                class="text-muted-foreground text-xs transition-all duration-300 ease-in-out max-h-[100px] opacity-100"
              >
                Test Description
              </p>
              <div
                class="mt-auto pt-2 flex items-center gap-2 z-30"
              >
                <div
                  class="inline-flex items-center rounded-full px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-secondary-foreground hover:bg-secondary/80 bg-background/80 border border-muted"
                >
                  test-tag
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="fixed bottom-4 right-4 flex items-center gap-2 z-50 bg-background/80 backdrop-blur-sm p-2 rounded-lg shadow-sm"
    />
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        class="fixed top-0 z-[100] flex max-h-screen w-full flex-col-reverse p-4 sm:bottom-0 sm:right-0 sm:top-auto sm:flex-col md:max-w-[420px]"
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
